File: old/actions/configure_haproxy_systemd.py
"""
Configure HAProxy systemd override action
"""
import logging
from cli import FileOps
from .base import Action
logger = logging.getLogger(__name__)

class ConfigureHaproxySystemdAction(Action):
    """Action to configure systemd override for HAProxy"""
    description = "haproxy systemd override"

    def execute(self) -> bool:
        """Configure systemd override for HAProxy"""
        if not self.ssh_service:
            logger.error("SSH service not initialized")
            return False
        mkdir_cmd = FileOps().mkdir("/etc/systemd/system/haproxy.service.d", parents=True)
        output, exit_code = self.ssh_service.execute(mkdir_cmd, sudo=True)
        if exit_code is not None and exit_code != 0:
            logger.error("create haproxy systemd override directory failed with exit code %s", exit_code)
            if output:
                logger.error("create haproxy systemd override directory output: %s", output.splitlines()[-1])
            return False
        override_content = (
            "[Service]\n"
            "Type=notify\n"
            "PrivateNetwork=no\n"
            "ProtectSystem=no\n"
            "ProtectHome=no\n"
            "ExecStart=\n"
            "ExecStart=/usr/sbin/haproxy -Ws -f $CONFIG -p $PIDFILE $EXTRAOPTS\n"
        )
        override_cmd = FileOps().write("/etc/systemd/system/haproxy.service.d/override.conf", override_content)
        output, exit_code = self.ssh_service.execute(override_cmd, sudo=True)
        if exit_code is not None and exit_code != 0:
            logger.error("write haproxy systemd override failed with exit code %s", exit_code)
            if output:
                logger.error("write haproxy systemd override output: %s", output.splitlines()[-1])
            return False
        reload_cmd = "systemctl daemon-reload"
        output, exit_code = self.ssh_service.execute(reload_cmd, sudo=True)
        if exit_code is not None and exit_code != 0:
            logger.error("reload systemd daemon failed with exit code %s", exit_code)
            if output:
                logger.error("reload systemd daemon output: %s", output.splitlines()[-1])
            return False
        return True

File: src/actions/configure_haproxy_systemd.go
package actions

import (
	"strings"
	"enva/cli"
	"enva/libs"
	"enva/services"
)

// ConfigureHaproxySystemdAction configures systemd override for HAProxy
type ConfigureHaproxySystemdAction struct {
	*BaseAction
}

func NewConfigureHaproxySystemdAction(sshService *services.SSHService, aptService *services.APTService, pctService *services.PCTService, containerID *string, cfg *libs.LabConfig, containerCfg *libs.ContainerConfig) Action {
	return &ConfigureHaproxySystemdAction{
		BaseAction: &BaseAction{
			SSHService:   sshService,
			APTService:   aptService,
			PCTService:   pctService,
			ContainerID: containerID,
			Cfg:          cfg,
			ContainerCfg: containerCfg,
		},
	}
}

func (a *ConfigureHaproxySystemdAction) Description() string {
	return "haproxy systemd override"
}

func (a *ConfigureHaproxySystemdAction) Execute() bool {
	if a.SSHService == nil {
		libs.GetLogger("configure_haproxy_systemd").Error("SSH service not initialized")
		return false
	}
	mkdirCmd := cli.NewFileOps().Mkdir("/etc/systemd/system/haproxy.service.d", true)
	output, exitCode := a.SSHService.Execute(mkdirCmd, nil, true) // sudo=True
	if exitCode != nil && *exitCode != 0 {
		libs.GetLogger("configure_haproxy_systemd").Error("create haproxy systemd override directory failed with exit code %d", *exitCode)
		if output != "" {
			lines := strings.Split(output, "\n")
			if len(lines) > 0 {
				libs.GetLogger("configure_haproxy_systemd").Error("create haproxy systemd override directory output: %s", lines[len(lines)-1])
			}
		}
		return false
	}
	overrideContent := `[Service]
Type=notify
PrivateNetwork=no
ProtectSystem=no
ProtectHome=no
ExecStart=
ExecStart=/usr/sbin/haproxy -Ws -f $CONFIG -p $PIDFILE $EXTRAOPTS
`
	overrideCmd := cli.NewFileOps().Write("/etc/systemd/system/haproxy.service.d/override.conf", overrideContent)
	output, exitCode = a.SSHService.Execute(overrideCmd, nil, true) // sudo=True
	if exitCode != nil && *exitCode != 0 {
		libs.GetLogger("configure_haproxy_systemd").Error("write haproxy systemd override failed with exit code %d", *exitCode)
		if output != "" {
			lines := strings.Split(output, "\n")
			if len(lines) > 0 {
				libs.GetLogger("configure_haproxy_systemd").Error("write haproxy systemd override output: %s", lines[len(lines)-1])
			}
		}
		return false
	}
	reloadCmd := "systemctl daemon-reload"
	output, exitCode = a.SSHService.Execute(reloadCmd, nil, true) // sudo=True
	if exitCode != nil && *exitCode != 0 {
		libs.GetLogger("configure_haproxy_systemd").Error("reload systemd daemon failed with exit code %d", *exitCode)
		if output != "" {
			lines := strings.Split(output, "\n")
			if len(lines) > 0 {
				libs.GetLogger("configure_haproxy_systemd").Error("reload systemd daemon output: %s", lines[len(lines)-1])
			}
		}
		return false
	}
	return true
}

